overlord/snapstate/handlers.go

```diff
@@ -1926,24 +1925,31 @@ func setMigrationFlagsinState(snapst *SnapState, snapsup *SnapSetup) {
 	}
 }
 
-// maybeRestart will schedule a reboot or restart as needed for the
-// just linked snap with info if it's a core or snapd or kernel snap.
-func (m *SnapManager) maybeRestart(t *state.Task, info *snap.Info, rebootInfo boot.RebootInfo) {
+// maybeRestart carries information to decide whether a restart of some form
+// is required.
+type maybeRestart struct {
+	info *snap.Info
+	boot.RebootInfo
+}
+
+// finishTaskWithMaybeRestart will set the final status for the task
+// and schedule a reboot or restart as needed for the just linked snap
+// with info if it's a core, snapd or kernel snap.
+func (m *SnapManager) finishTaskWithMaybeRestart(t *state.Task, status state.Status, maybeRst maybeRestart) error {
 	// Don't restart when preseeding - we will switch to new snapd on
 	// first boot.
 	if m.preseed {
-		return
+		return nil
 	}
 
 	st := t.State()
 
-	if rebootInfo.RebootRequired {
+	if maybeRst.RebootRequired {
 		t.Logf("Requested system restart.")
-		RestartSystem(t, &rebootInfo)
-		return
+		return FinishTaskWithRestart(t, status, restart.RestartSystem, &maybeRst.RebootInfo)
 	}
 
-	typ := info.Type()
+	typ := maybeRst.info.Type()
```

Comment:
Again, not a strong opinion, but here that we have more code between the two `return` statement, I'd use an `else`. It may also be that it's possible to refactor the code in a way that the call to `FinishTaskWithRestart()` appears just once at the end. Not sure it's worth, though.